pkg/agent/controller/noderoute/node_route_controller.go

```diff
@@ -471,43 +496,66 @@ func (c *Controller) addNodeRoute(nodeName string, node *corev1.Node) error {
 		}
 		peerGatewayIP := ip.NextIP(peerPodCIDRAddr)
 		peerConfig[peerPodCIDR] = peerGatewayIP
-		podCIDRs = append(podCIDRs, peerPodCIDR)
+		peerPodCIDRs = append(peerPodCIDRs, peerPodCIDR)
 	}
 
-	ipsecTunOFPort := int32(0)
-	if c.networkConfig.EnableIPSecTunnel {
-		// Create a separate tunnel port for the Node, as OVS IPSec monitor needs to
-		// read PSK and remote IP from the Node's tunnel interface to create IPSec
-		// security policies.
-		if ipsecTunOFPort, err = c.createIPSecTunnelPort(nodeName, peerNodeIP); err != nil {
-			return err
-		}
+	peerNodeIPs, err := k8s.GetNodeAddrs(node)
+	if err != nil {
+		klog.Errorf("Failed to retrieve IP addresses of Node %s: %v", nodeName, err)
+		return nil
 	}
 
-	err = c.ofClient.InstallNodeFlows(
-		nodeName,
-		peerConfig,
-		peerNodeIP,
-		uint32(ipsecTunOFPort),
-		peerNodeMAC)
-	if err != nil {
-		return fmt.Errorf("failed to install flows to Node %s: %v", nodeName, err)
+	for _, peerNodeIP := range peerNodeIPs {
+		ipsecTunOFPort := int32(0)
+		if c.networkConfig.EnableIPSecTunnel {
+			// Create a separate tunnel port for the Node, as OVS IPSec monitor needs to
+			// read PSK and remote IP from the Node's tunnel interface to create IPSec
+			// security policies.
+			if ipsecTunOFPort, err = c.createIPSecTunnelPort(nodeName, peerNodeIP); err != nil {
+				return err
+			}
+		}
+
+		err = c.ofClient.InstallNodeFlows(
+			nodeName,
+			peerConfig,
+			peerNodeIP,
+			uint32(ipsecTunOFPort),
+			peerNodeMAC)
+		if err != nil {
+			return fmt.Errorf("failed to install flows to Node %s: %v", nodeName, err)
+		}
 	}
 
 	var peerGatewayIPs []net.IP
+	var peerNodeIPv4, peerNodeIPv6 net.IP
+	for _, ip := range peerNodeIPs {
```

Comment:
I start to think could it make code simpler to define a struct which has two fields to save v4 and v6 IP respectively, and the struct should be initialized at the time the IPs are discovered, and used by call code that handles dual stack IPs, then all other code needs not to check v4/v6 again.

Do you think that can help?

Reply:
Good idea. Updated.